client/src/erlang/receiver.zig

```diff
@@ -0,0 +1,311 @@
+        ei.ei_get_type(deserializer.buf.buff, deserializer.index, &typ, &_v),
+    );
+    const enum_type = std.meta.Tag(T);
+    if (typ == ei.ERL_ATOM_EXT) {
+        const tuple_name = try receive_enum(enum_type, @typeInfo(enum_type).Enum, allocator, deserializer);
+        switch (tuple_name) {
+            inline else => |name| {
+                inline for (item.fields) |field| {
+                    if (field.type == void and comptime std.mem.eql(
+                        u8,
+                        field.name,
+                        @tagName(name),
+                    )) {
+                        value = name;
+                        break;
+                    }
+                } else return error.invalid_union_tag;
+            },
+        }
+        return value;
+    } else {
+        try erl.validate(
+            error.decoding_tuple,
+            ei.ei_decode_tuple_header(deserializer.buf.buff, deserializer.index, &arity),
+        );
+        if (arity != 2) {
+            return error.wrong_arity_for_tuple;
+        }
+        const tuple_name = try receive_enum(enum_type, @typeInfo(enum_type).Enum, allocator, deserializer);
+        switch (tuple_name) {
+            inline else => |name| {
+                inline for (item.fields) |field| {
+                    if (field.type != void and comptime std.mem.eql(
+                        u8,
+                        field.name,
+                        @tagName(name),
+                    )) {
+                        const tuple_value = try internal_receive_message(field.type, allocator, deserializer);
+                        value = @unionInit(T, field.name, tuple_value);
+                        break;
+                    }
+                } else return error.failed_to_receive_payload;
+            },
+        }
+    }
+    return error.unknown_tuple_tag;
+}
+
+inline fn receive_pointer(comptime T: type, comptime item: std.builtin.Type.Pointer, allocator: std.mem.Allocator, deserializer: Deserializer) !T {
+    var value: T = undefined;
+    if (item.size != .Slice)
+        return error.unsupported_pointer_type;
+    var size: i32 = 0;
+    try erl.validate(
+        error.decoding_list_in_pointer_1,
+        ei.ei_decode_list_header(deserializer.buf.buff, deserializer.index, &size),
+    );
+    const has_sentinel = item.sentinel != null;
+    if (size == 0 and !has_sentinel) {
+        value = &.{};
+    } else {
+        const usize_size: u32 = @bitCast(size);
+        const slice_buffer = if (has_sentinel)
+            try allocator.allocSentinel(
+                item.child,
+                usize_size,
+                item.sentinel.?,
+            )
+        else
+            try allocator.alloc(
+                item.child,
+                usize_size,
+            );
+        errdefer allocator.free(slice_buffer);
+        for (slice_buffer) |*elem| {
+            elem.* = try internal_receive_message(item.child, allocator, deserializer);
+        }
```

Comment:
We should do something about possibly having to deallocate the child types in this slice if one of the intermediate calls to `internal_receive_message` fails.

Reply:
Added as a TODO.